### PR TITLE
porting of ChooseRadioAccessTechnology from MKRNB

### PR DESCRIPTION
**Pull request by Rocketct**
porting of ChooseRadioAccessTechnology from MKRNB library,

Tested and works!

cc:/ @sandeepmistry could you double check and check also the coments?

Datasheet reference for allowed URAT combination(https://www.u-blox.com/sites/default/files/u-blox-CEL_ATCommands_%28UBX-13002752%29.pdf page 101):

urat: 0 - 2G only
urat :2 - 3G only
urat 1,0 - 2G/3G, 2G preferred
urat 1,2 -  2G/3G, 3G preferred (default)

### CODE REVIEW

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino [hunk 1]
```diff
@@ -0,0 +1,126 @@
+
+/*
+  Radio Access Technology selection for Arduino MKR GSM 1400
+  This sketch allows you to select your preferred Radio Access
+  Technology (RAT).
+  You can choose among 2G, 3G(UMTS) or a combination of both.
+  Selecting JUST ONE technology will speed up the modem's network
+  registration A LOT!
+  The chosen configuration will be saved to modem's internal memory
+  and will be preserved through MKR GSM 1400 sketch uploads.
```
Comment (sandeepmistry) [inline examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: The spacing seems different from the original here: https://github.com/arduino-libraries/MKRNB/blob/master/examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  - MKR GSM 1400 board
+  - Antenna
+  - SIM card
+  Created 19 August 2019
+  by Riccardo Rizzo
+*/
+
+#include <MKRGSM.h>
+
+GSM gms(true);
```
Comment (sandeepmistry) [inline examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: Is this line needed?

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  MODEM.begin();
+  while (!MODEM.noop());
+
+  for (int i = 0; i < 80; i++) Serial.print("*");
+  Serial.println();
+  Serial.println("This sketch allows you to select your preferred");
+  Serial.println("GSM Radio Access Technology (RAT).");
+  Serial.println();
+  Serial.println("You can choose among 2G, UMTS(3G) or a combination of both.");
```
Comment (sandeepmistry) [inline examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: Maybe 3G/UTMS here?

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  for (int i = 0; i < 80; i++) Serial.print("*");
+
+  Serial.println();
+  Serial.println();
+  Serial.println("Please choose your Radio Access Technology:");
+  Serial.println();
+  Serial.println("    0 - 2G only");
+  Serial.println("    1 - 3G only");
+  Serial.println("    2 - 2G preferred, 3G as failover (default)");
+  Serial.println("    3 - 3G IoT preferred, 2G as failover");
```
Comment (sandeepmistry) [inline examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: What's 3G IoT :)

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  Serial.println("In order to change the RAT, you will need to run this sketch again.");
+  for (int i = 0; i < 80; i++) Serial.print("*");
+
+  Serial.println();
+  Serial.println();
+  Serial.println("Please choose your Radio Access Technology:");
+  Serial.println();
+  Serial.println("    0 - 2G only");
+  Serial.println("    1 - 3G only");
+  Serial.println("    2 - 2G preferred, 3G as failover (default)");
```
Comment (sandeepmistry) [inline examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: From the AT command reference module the default is UART= 1 = GSM / UMTS (dual mode)

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+  switch (urat) {
+    case 0:
+      uratChoice = "0";
+      break;
+    case 1:
+      uratChoice = "2";
+      break;
+    case 2:
+      uratChoice = "1,0";
```
Comment (sandeepmistry) [inline examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: Should we make this 1,2?

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino (hunk 1 above)
Comment (Rocketct) [reply examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: changed!

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  Created 19 August 2019
+  by Riccardo Rizzo
+
+*/
+
+#include <MKRGSM.h>
+
+
+void setup() {
+  Serial.begin(921600);
```
Comment (sandeepmistry) [inline examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: I think the baud should be 9600 here.

File: examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino [hunk 8]
```diff
@@ -0,0 +1,133 @@
+/*
+  Radio Access Technology selection for Arduino MKR GSM 1400
+
+  This sketch allows you to select your preferred Radio Access
+  Technology (RAT).
+
+  You can choose among 2G, 3G/UTMS or a combination of both.
+
+  Selecting JUST ONE technology will speed up the modem's network
```
Comment (sandeepmistry) [inline examples/Tools/ChooseRadioAccessTechnology/ChooseRadioAccessTechnology.ino]: I think we can remove this line, might apply to MKR NB 1500 only.

**Comment by sandeepmistry** [comment]
Tested this sketch and things look good. Please see my two comments and we are good to go!

**Comment by terrasmart** [comment]
This will be very useful, thank you.

If I may ask, the raw sequence of AT commands extracted from the code is:

AT+COPS=2
AT+URAT=2 (if you want 3G only)
AT+CFUN=16

correct ?

**Comment by Rocketct** [comment]
HI @terrasmart yes is correct!!

**Pull request merged**